File: packages/php/src/Seguranca/Ataques/AtaqueTokensMalformados.php
<?php

namespace Application\Seguranca\Ataques;

use Application\Seguranca\AlvoCriptografico;
use Application\Seguranca\AtaqueInterface;
use Application\Seguranca\ResultadoAtaque;

/**
 * Robustez do parser de tokens: qualquer entrada que não seja um token
 * íntegro e bem formado DEVE ser rejeitada com exceção. Um decrypt() que
 * devolve lixo em vez de lançar (ou que aceita truncamentos/bytes extras)
 * é uma porta pra bugs de validação e "token smuggling".
 */
class AtaqueTokensMalformados implements AtaqueInterface
{
    public function nome(): string
    {
        return 'Tokens malformados (fuzzing de entrada)';
    }

    public function executar(AlvoCriptografico $alvo): ResultadoAtaque
    {
        $token = $alvo->encrypt('MENSAGEM_VALIDA_PARA_TESTE');
        $prefixo = $alvo->prefixo();
        $corpo = substr($token, strlen($prefixo));

        $casos = [
            'vazio' => '',
            'só prefixo' => $prefixo,
            'prefixo errado' => 'XXX' . $corpo,
            'base64 inválido' => $prefixo . '!!!@@@###',
            'bytes extras no fim' => $token . 'AAAA',
            'bytes extras no início' => 'AAAA' . $token,
        ];
        for ($len = 1; $len < strlen($token); $len++) {
            $casos["truncado em $len"] = substr($token, 0, $len);
        }
        for ($i = 0; $i < 20; $i++) {
            $casos["lixo aleatório $i"] = $prefixo . bin2hex(random_bytes(16));
        }

        $aceitos = [];
        foreach ($casos as $nome => $t) {
            try {
                $r = $alvo->decrypt($t);
                $aceitos[] = sprintf('%s -> aceito (retornou %d bytes)', $nome, strlen($r));
            } catch (\Throwable $e) {
                // comportamento esperado
            }
        }

        if (!empty($aceitos)) {
            return new ResultadoAtaque(
                $this->nome(),
                vulneravel: true,
                severidade: 'critica',
                detalhes: count($aceitos) . ' entrada(s) malformada(s) foram ACEITAS em vez de rejeitadas',
                dados: ['exemplos' => array_slice($aceitos, 0, 10)],
            );
        }

        return new ResultadoAtaque(
            $this->nome(),
            vulneravel: false,
            severidade: 'info',
            detalhes: 'Todas as ' . count($casos) . ' entradas malformadas foram rejeitadas',
        );
    }
}

File: packages/rust/src/seguranca/ataques/ataque_tokens_malformados.rs
//! Robustez do parser de tokens: qualquer entrada que não seja um token
//! íntegro e bem formado DEVE ser rejeitada com exceção. Um decrypt() que
//! devolve lixo em vez de lançar (ou que aceita truncamentos/bytes extras)
//! é uma porta pra bugs de validação e "token smuggling".

use crate::seguranca::alvo_criptografico::AlvoCriptografico;
use crate::seguranca::ataque::{Ataque, ErroAtaque};
use crate::seguranca::resultado_ataque::{ResultadoAtaque, Severidade};
use crate::seguranca::util::{hex_encode, random_bytes};

pub struct AtaqueTokensMalformados;

impl Ataque for AtaqueTokensMalformados {
    fn nome(&self) -> String {
        "Tokens malformados (fuzzing de entrada)".to_string()
    }

    fn executar(&self, alvo: &dyn AlvoCriptografico) -> Result<ResultadoAtaque, ErroAtaque> {
        let token = alvo
            .encrypt(b"MENSAGEM_VALIDA_PARA_TESTE")
            .map_err(|e| ErroAtaque::Falha(e.to_string()))?;
        let prefixo = alvo.prefixo();
        let corpo = &token[prefixo.len()..];

        let mut casos: Vec<(String, String)> = Vec::new();
        casos.push(("vazio".to_string(), String::new()));
        casos.push(("só prefixo".to_string(), prefixo.to_string()));
        casos.push(("prefixo errado".to_string(), format!("XXX{}", corpo)));
        casos.push((
            "base64 inválido".to_string(),
            format!("{}!!!@@@###", prefixo),
        ));
        casos.push(("bytes extras no fim".to_string(), format!("{}AAAA", token)));
        casos.push((
            "bytes extras no início".to_string(),
            format!("AAAA{}", token),
        ));

        for len in 1..token.len() {
            casos.push((
                format!("truncado em {}", len),
                token[..len].to_string(),
            ));
        }
        for i in 0..20 {
            casos.push((
                format!("lixo aleatório {}", i),
                format!("{}{}", prefixo, hex_encode(&random_bytes(16))),
            ));
        }

        let mut aceitos: Vec<String> = Vec::new();
        for (nome, t) in &casos {
            if let Ok(r) = alvo.decrypt(t) {
                aceitos.push(format!(
                    "{} -> aceito (retornou {} bytes)",
                    nome,
                    r.len()
                ));
            }
        }

        if !aceitos.is_empty() {
            return Ok(ResultadoAtaque::new(
                &self.nome(),
                true,
                Severidade::Critica,
                &format!(
                    "{} entrada(s) malformada(s) foram ACEITAS em vez de rejeitadas",
                    aceitos.len()
                ),
            ));
        }

        Ok(ResultadoAtaque::new(
            &self.nome(),
            false,
            Severidade::Info,
            &format!(
                "Todas as {} entradas malformadas foram rejeitadas",
                casos.len()
            ),
        ))
    }
}
